### DISCUSSION TITLE
JSON.parse error when writing data to MySQL database

### DISCUSSION THREAD

**Discussion by d-bohn** [jspsych/jsPsych#1124] [Q&A]
Hello,

I am trying to write my data to a MySQL database following the suggested instructions on the [website](https://www.jspsych.org/overview/data/#storing-data-permanently-in-a-mysql-database). My `php` files and `js` file are identical to the ones on the website, however, when I run the experiment on the server, I get the following error:

```
Uncaught SyntaxError: JSON.parse: unexpected character at line 1 column 1 of the JSON data
```

A few other things I should note: First, my experiment is bundled into a single file `app.js` (though I don't think this should matter). Second, the MySQL database I am trying to write to is hosted on a different server, however, the configuration options in `database_config.php` reflect this, and I have confirmed that I can remote access the database server. Lastly, I have already created a table on the database server with the columns of interest from the experiment output (as per the comments on the website detailing how to save data to MySQL databases.

I am new to `php`, so any help would be appreciated! Thanks!

**Comment by d-bohn** [comment]
OK, so after much debugging, I can now write individual trials to the remote database using the saveData MySQL function with `on_finish` after each trial. However, if I try to save all of the data at the end of the experiment I get a similar `JSON.parse` error:

```
Uncaught SyntaxError: JSON.parse: unexpected character at line 1 column 31 of the JSON data
```

The weird thing is that it will write one row to the database (the very first trial), but then nothing after that.

EDIT:
I should note that there are not even 31 variables being written out to the database, or in the experiment at all (confirmed with `jsPsych.data.displayData("csv")` at the end of the experiment).

**Comment by d-bohn** [comment]
I believe I have it figured out. The issue was a combination of the database expecting an integer value and for some reason the one of the `html-keyboard-response` trials recording keyboard responses. Still not sure where "column 31" came from, though.

**Reply by holgerMitterer** [reply]
I also got the parse error message, and realized that it is actually not caused by writing to the databse, but by the feedback returned to javascript. I was using the code as it is on the website, (I learned that the 200 status means the query has finished)
//from the save2Database function in javascript 
  if(xhr.status == 200){
      var response = JSON.parse(xhr.responseText);
      console.log(response.success);
    }

and changed it into this, and now thinks work
  if(xhr.status == 200){
      var response = xhr.responseText;
      console.log(response);
    }

the output io console.log from this is {"success": true}, and I don't understand why this is not JSON parsable, but as long as it works ....
